File: samples/rust_samples/bench_primes_sieve.rs
mod common;

use common::{read_line, repeat, write_report, ProfileInfo};

static SIEVE_INFO: ProfileInfo = ProfileInfo::new();

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Primality {
    Prime,
    Composite,
}

fn sieve(limit: u64) -> Vec<Primality> {
    SIEVE_INFO.record_call(|| {
        let mut arr = vec![Primality::Prime; limit as usize];
        arr[0] = Primality::Composite;
        arr[1] = Primality::Composite;

        for n in 2..limit {
            if arr[n as usize] == Primality::Prime {
                let mut i = 2 * n;
                while i < limit {
                    arr[i as usize] = Primality::Composite;
                    i += n;
                }
            }
        }

        arr
    })
}

fn main() {
    match (read_line().parse::<u64>(), read_line().parse::<u64>()) {
        (Ok(iters), Ok(limit)) => {
            if let Some(primes) = repeat(iters, || sieve(limit)) {
                for i in 0..limit {
                    match primes[i as usize] {
                        Primality::Prime => println!("{}", i),
                        Primality::Composite => {}
                    }
                }
            }
        }

        (_, _) => eprintln!("Please enter two positive integers (an iteration count and a limit)"),
    }

    write_report(&SIEVE_INFO);
}

File: samples/haskell_samples/BenchPrimesSieve.hs
{-# language TupleSections #-}

module BenchPrimesSieve where

import Control.Monad (forM_)
import Data.List (foldl')
import Data.IntMap.Strict (IntMap)
import qualified Data.IntMap.Strict as IntMap
import Text.Read (readMaybe)

import BenchCommon (runIters)

data Primality = Prime | Composite deriving (Eq, Show)

{-# scc sieve #-}
sieve :: Int -> () -> IntMap Primality
sieve limit () =
  let
    initArr =
      IntMap.fromList $ [(0, Composite), (1, Composite)] ++ map (, Prime) [2..limit - 1]

    markMultiples arr n =
      case arr IntMap.! n of
        Prime ->
          foldl' (\new i -> IntMap.adjust (const Composite) i new) arr $
          takeWhile (< limit) $
          map (* n) $
          [2..]

        Composite ->
          arr
  in
    foldl' markMultiples initArr [2..limit - 1]

main :: IO ()
main = do
  itersStr <- getLine
  limitStr <- getLine
  case (readMaybe itersStr, readMaybe limitStr) of
    (Just iters, Just limit) ->
      case runIters (sieve limit) iters of
        Just primes ->
          forM_ (IntMap.toAscList primes) $ \(i, status) ->
            case status of
              Prime -> print i
              Composite -> return ()

        Nothing ->
          return ()

    (_, _) ->
      putStrLn "Please enter two positive integers (an iteration count and a limit)"
